### ISSUE TITLE
Allow callbacks to emit ragged (variable-length per-sample) tensors via Emit.sample_lengths

### ISSUE THREAD

**Issue by AlliedToasters** [AlliedToasters/fpwap#65]
## Motivation

When extracting activations for downstream tasks like probing, the consumer typically only cares about a small subset of token positions per sample — e.g., assistant-role tokens in a dialogue corpus, where system+user+padding positions are write-once-then-discard. On a typical dialogue corpus this is **5-30% of the dense emit**.

Today, fpwap callbacks have no way to express "emit only these positions per sample." A callback can return a smaller `Emit(tensor=...)` but `engine.py:1620-1644` interprets it as fixed-shape and **left-pads with zeros up to `seq_len`** before handing it to storage. So even a callback that pre-trims gets re-padded back to dense, and `MemmapBackend` writes max-seq-padded emits regardless.

For a 70B / 14k-sample / max_seq=3135 sweep at one detect layer, this is the difference between:
- Dense emit: `[14131, 3135, 8192]` × bf16 ≈ **57 GiB on disk**
- Ragged emit (5% kept): `[~750k_tokens, 8192]` × bf16 ≈ **12 GiB on disk**

The 5x reduction directly relaxes the page-cache budget that K-sweep packing has to respect.

## Observed

`engine.py:1627-1643`:

```python
if isinstance(result, Emit):
    if self.storage is not None:
        emit_tensor = result.tensor
        if (
            emit_tensor.dim() >= 2
            and emit_tensor.shape[1] < self.seq_len
        ):
            gap = self.seq_len - emit_tensor.shape[1]
            pad_shape = (emit_tensor.shape[0], gap, *emit_tensor.shape[2:])
            z = torch.zeros(...)
            emit_tensor = torch.cat([z, emit_tensor], dim=1)
        self.storage.write_emit(layer_idx, hook, sample_ids, emit_tensor)
```

There is no escape hatch. A callback returning `[total_kept_toks, hidden]` (genuine ragged, 2D) hits this same branch — `shape[1]` is misinterpreted as a seq dim, padding goes along the wrong axis, silent corruption or crash.

## Proposed change

Augment `Emit` with one optional field:

```python
@dataclass
class Emit:
    tensor: Tensor
    sample_lengths: Tensor | None = None   # NEW: [N_samples] int counts
```

Semantics:
- `sample_lengths is None` → existing behavior (dense, `[batch, seq, hidden]`, padded if short).
- `sample_lengths is not None` → `tensor` is interpreted as `[sum(sample_lengths), *trailing_shape]` (ragged, sample-dim concatenated). Engine **skips the left-pad block** and dispatches to ragged storage.

Engine change: one `if result.sample_lengths is not None` branch around the pad block.

Storage backend change (`MemmapBackend`):
- Add a ragged layout: store flat tensor `[total_kept_across_sweep, *trailing]` plus a per-sample offset/count sidecar `[N_samples + 1]`.
- `write_emit` gains an optional `sample_lengths` kwarg matching the engine signal.
- `read_all` returns either dense tensor (existing) or `(flat_tensor, offsets)` / `RaggedTensor` (new), based on what was written. Sidecar JSON records which layout each shard uses.

## Why this scope is right

- **Mechanism, not policy.** fpwap learns one concept: "variable-length per sample is a thing." It doesn't learn what a mask/dialogue/role/token is. Those stay in user code.
- **Backwards compatible.** Every existing user keeps working — `sample_lengths=None` is default and matches today's contract.
- **No new abstractions.** Callbacks still return `Emit`. Callers still call `result.activations(...)`. The dataclass just gets richer.
- **Composable with existing fixes.** `posix_fadvise(DONTNEED)`, async D2H staging, bucketed reassembly all work on the flat ragged tensor the same way they work on the dense one.

## Caller-side helper (out of scope for fpwap)

Downstream wrapper code provides the mask logic:

```python
def mask_to_ragged_emit(tensor: [B, S, H], mask: [B, S]) -> Emit:
    return Emit(
        tensor=tensor[mask],          # [total_kept, H]
        sample_lengths=mask.sum(dim=1),
    )
```

This stays in the user's library — fpwap never sees a mask.

## Relationship to other issues

- #47 (bucketed write shape-mismatch) is also about the storage backend not handling variable-shape emits. The ragged layout proposed here is a clean superset: per-sample variable length subsumes per-bucket variable seq_len. Worth landing the ragged storage path first, then #47's seq-dim variability becomes a special case.
- #52 (read materializes full emit) is partially mooted by ragged: a 5x smaller emit is much less likely to OOM on readback even if the materialization stays the way it is post-#52.

## Repro / motivating use case

70B Llama-3.3 deception-probe extraction, full paper corpus (14k samples, max_seq 3135, single detect layer). Detection mask covers ~5% of token positions (assistant-only). Dense emit costs 57 GiB on disk and dominates the K-sweep cache budget. With ragged emit, single-sweep K=1 becomes feasible on a 128 GiB host.
